src/librustc/traits/specialize/specialization_graph.rs

```diff
@@ -291,8 +335,52 @@ impl<'a, 'gcx, 'tcx> Graph {
 
     /// The parent of a given impl, which is the def id of the trait when the
     /// impl is a "specialization root".
-    pub fn parent(&self, child: DefId) -> DefId {
-        *self.parent.get(&child).unwrap()
+    pub fn parent(&self, child: DefId) -> Vec<DefId> {
+        self.parent.get(&child).unwrap().clone()
+    }
+
+    pub fn build_graph(&self) -> (DataStructuresGraph<DefId, String>, DefIdMap<NodeIndex>) {
```

Comment:
Comment would be great. What are these two things that are being returned, anyway?

Reply:
For example, worth documenting that in the graph, we have edges *from* children *to* parents

Reply:
it might be nice to replace it, yes, though maybe that should be a separate PR

Reply:
i.e., it'd be nice to do that refactoring *first* without changing semantics